nix/devtools.nix

```diff
@@ -1,49 +1,96 @@
-{ writeShellScriptBin
-, buildEnv
+{ buildEnv
+, cabal-install
+, checkedShellScript
+, entr
 , git
 , hlint
 , nixpkgs-fmt
 , silver-searcher
 , stylish-haskell
+, tests
 }:
 let
   style =
-    writeShellScriptBin "postgrest-style"
+    checkedShellScript "postgrest-style"
       ''
-        set -euo pipefail
-
         rootdir="$(${git}/bin/git rev-parse --show-toplevel)"
 
         # Format Nix files
         ${nixpkgs-fmt}/bin/nixpkgs-fmt "$rootdir" > /dev/null 2> /dev/null
 
         # Format Haskell files
-        ${silver-searcher}/bin/ag -l -g '\.l?hs$' "$rootdir" \
+        ${silver-searcher}/bin/ag -l -g '\.l?hs$' . "$rootdir" \
           | xargs ${stylish-haskell}/bin/stylish-haskell -i
       '';
 
-  check =
-    writeShellScriptBin "postgrest-style-check"
+  # Script to check whether any uncommited changes result from postgrest-style
+  styleCheck =
+    checkedShellScript "postgrest-style-check"
       ''
-        set -euo pipefail
-
-        ${style}/bin/${style.name}
+        ${style}
 
         ${git}/bin/git diff-index --exit-code HEAD -- '*.hs' '*.lhs' '*.nix'
       '';
+
   lint =
-    writeShellScriptBin "postgrest-lint"
+    checkedShellScript "postgrest-lint"
       ''
-        set -euo pipefail
-
         rootdir="$(${git}/bin/git rev-parse --show-toplevel)"
 
         # Lint Haskell files
         ${silver-searcher}/bin/ag -l -g '\.l?hs$' "$rootdir" \
           | xargs ${hlint}/bin/hlint -X QuasiQuotes -X NoPatternSynonyms
       '';
+
+  watch =
+    checkedShellScript "postgrest-watch"
+      ''
+        rootdir="$(${git}/bin/git rev-parse --show-toplevel)"
+
+        while true; do
+          ${silver-searcher}/bin/ag -l . "$rootdir" | ${entr}/bin/entr -rd "$@"
+        done
+      '';
+
+  pushCachix =
+    checkedShellScript "postgrest-push-cachix"
+      ''
+        nix-store -qR --include-outputs "$(nix-instantiate)" \
+          | cachix push postgrest
+      '';
+
+  build =
+    checkedShellScript "postgrest-build"
+      ''exec ${cabal-install}/bin/cabal v2-build "$@"'';
+
+  run =
+    checkedShellScript "postgrest-run"
+      ''exec ${cabal-install}/bin/cabal v2-run postgrest -- "$@"'';
+
+  clean =
+    checkedShellScript "postgrest-clean"
+      ''
+        ${cabal-install}/bin/cabal v2-clean
+      '';
+
+  check =
+    checkedShellScript "postgrest-check"
+      ''
+        ${style}
+        ${tests}/bin/postgrest-test-spec-all
```

Comment:
Can we flip those two around, and also add `${lint}`?

I think 1. test-spec, 2. lint, 3. style would be the best order. I first want to make sure that the specs pass, then make the linter happy once the code does what it should do, and then let style do some whitespace stuff.

But if the linter fails before the specs are run, doing a `postgrest-watch postgrest-check` might be annoying.

Reply:
Yes, good idea! I'll also have to add a note that io and memory tests need to be run separately, as they depend on the postgrest executable as built by Nix (at least for now)

Reply:
Yeah, I just looked into that as well. Can we not just add a `exec ${cabal-install}/bin/cabal v2-exec` before the memory and io tests scripts? This should already fix this, I think. I remember doing that back in the stack days, with stack exec.

Reply:
Yes, that should work! The profiled build for memory is a bit tricky though, so let's do that in a separate PR.